app/controllers/pensions/unauthorisedPayments/DidYouPayNonUkTaxController.scala

```diff
@@ -0,0 +1,116 @@
+import play.api.mvc.{Action, AnyContent, MessagesControllerComponents}
+import services.PensionSessionService
+import uk.gov.hmrc.play.bootstrap.frontend.controller.FrontendController
+import utils.Clock
+import views.html.pensions.unauthorisedPayments.DidYouPayNonUKTaxSurchargeView
+
+import javax.inject.Inject
+import scala.concurrent.Future
+
+class DidYouPayNonUkTaxController @Inject()(implicit val cc: MessagesControllerComponents,
+                                            authAction: AuthorisedAction,
+                                            didYouPayNonUkTaxSurchargeView: DidYouPayNonUKTaxSurchargeView,
+                                            appConfig: AppConfig,
+                                            pensionSessionService: PensionSessionService,
+                                            errorHandler: ErrorHandler,
+                                            clock: Clock) extends FrontendController(cc) with I18nSupport {
+
+
+  def form: Form[(Boolean, Option[BigDecimal])] = RadioButtonAmountForm.radioButtonAndAmountForm(
+    missingInputError = s"unauthorisedPayments.didYouPayNonUkTax.error.noEntry",
+    emptyFieldKey = s"unauthorisedPayments.didYouPayNonUkTax.error.Amount.noEntry",
+    wrongFormatKey = s"unauthorisedPayments.didYouPayNonUkTax.error.Amount.incorrectFormat",
+    exceedsMaxAmountKey = s"unauthorisedPayments.didYouPayNonUkTax.error.Amount.overMaximum"
+  )
+
+
+  def show(taxYear: Int): Action[AnyContent] = authAction.async { implicit request => {
+    pensionSessionService.getPensionsSessionDataResult(taxYear, request.user) {
+      case Some(data) =>
+        if (data.pensions.unauthorisedPayments.surchargeAmount.isDefined) {
+          data.pensions.unauthorisedPayments.surchargeTaxAmount match {
+            case Some(value) if value == 0 =>
+              Future.successful(Ok(didYouPayNonUkTaxSurchargeView(form.fill((false, Some(value))), taxYear)))
+
+            case Some(value) =>
+              Future.successful(Ok(didYouPayNonUkTaxSurchargeView(form.fill((true, Some(value))), taxYear)))
+
+            case None =>
+              Future.successful(Ok(didYouPayNonUkTaxSurchargeView(form, taxYear)))
+          }
+        } else {
+          //todo redirect to unauthorised cya page
+          Future.successful(Ok(didYouPayNonUkTaxSurchargeView(form, taxYear)))
+        }
+      case _ =>
+        //todo - redirect to unauthorised cya page
+        Future.successful(Redirect(PensionsSummaryController.show(taxYear)))
+    }
+  }
+
+  }
+
+
+  def submit(taxYear: Int): Action[AnyContent] = authAction.async { implicit request => {
+    pensionSessionService.getPensionsSessionDataResult(taxYear, request.user) {
+      case Some(data) =>
+        if (data.pensions.unauthorisedPayments.surchargeAmount.isDefined) {
+          form.bindFromRequest.fold(
+            formWithErrors => Future.successful(BadRequest(didYouPayNonUkTaxSurchargeView(formWithErrors, taxYear))),
+            amounts => {
```

Comment:
You could use a case here, with both parts of the tuple.

```
case (yesSelected, amountOpt) => {
    val pensionsCYAModel: PensionsCYAModel = data.pensions
    val viewModel: UnauthorisedPaymentsViewModel = pensionsCYAModel.unauthorisedPayments
    val updatedCyaModel: PensionsCYAModel = {
      pensionsCYAModel.copy(
        unauthorisedPayments = viewModel.copy(
          surchargeTaxAmountQuestion = Some(yesSelected),
          surchargeTaxAmount = if (yesSelected) amountOpt else Some(0)
        )
      )
    }
    pensionSessionService.createOrUpdateSessionData(request.user,
      updatedCyaModel, taxYear, data.isPriorSubmission)(errorHandler.internalServerError()) {
      //TODO: next page (were any of the unauthorised payments from a uk pension scheme page)
      Redirect(DidYouPayNonUkTaxController.show(taxYear))
    }
  }
```
That way, we don't have to reference the tuple parts by index.

Reply:
Sorry - I should have compiled that. Ignore for now.

Reply:
Not sure why but the type isn't resolving on my machine it says `yesSelected` is of type `Any`

Reply:
This compiles, and all tests pass:

```
            amounts => {
              val pensionsCYAModel: PensionsCYAModel = data.pensions
              val viewModel: UnauthorisedPaymentsViewModel = pensionsCYAModel.unauthorisedPayments
              val updatedCyaModel: PensionsCYAModel = amounts match {
                case (yesSelected, amountOpt) =>
                  pensionsCYAModel.copy(
                    unauthorisedPayments = viewModel.copy(
                      surchargeTaxAmountQuestion = Some(yesSelected),
                      surchargeTaxAmount = if (yesSelected) amountOpt else Some(0)
                    )
                  )
              }
              pensionSessionService.createOrUpdateSessionData(request.user,
                updatedCyaModel, taxYear, data.isPriorSubmission)(errorHandler.internalServerError()) {
                //TODO: next page (were any of the unauthorised payments from a uk pension scheme page)
                Redirect(DidYouPayNonUkTaxController.show(taxYear))
              }
            }
```

Reply:
It doesn't. The second one is correct.

Reply:
Confirmed.